chain-signatures/contract/src/lib.rs

```diff
@@ -320,39 +387,50 @@ impl VersionedMpcContract {
                             &data_id,
                             &serde_json::to_vec(&response).unwrap(),
                         );
+                        Ok(())
                     } else {
-                        env::panic_str(
-                            "this sign request was removed from pending requests: timed out or completed.",
-                        )
+                        Err(MpcContractError::SignatureError(
+                            SignatureError::RequestNotInPending,
+                        ))
                     }
                 }
             }
         } else {
-            env::panic_str("protocol is not in a running state");
+            Err(MpcContractError::ProtocolStateError(
+                ProtocolStateError::NotExpectedState("running".to_string()),
+            ))
         }
     }
 
     /// This is the root public key combined from all the public keys of the participants.
-    pub fn public_key(&self) -> PublicKey {
+    #[handle_result]
+    pub fn public_key(&self) -> Result<PublicKey, MpcContractError> {
         match self.state() {
-            ProtocolContractState::Running(state) => state.public_key.clone(),
-            ProtocolContractState::Resharing(state) => state.public_key.clone(),
-            _ => env::panic_str("public key not available (protocol is not running or resharing)"),
+            ProtocolContractState::Running(state) => Ok(state.public_key.clone()),
+            ProtocolContractState::Resharing(state) => Ok(state.public_key.clone()),
+            _ => Err(MpcContractError::ProtocolStateError(
+                ProtocolStateError::NotExpectedState("running or resharing".to_string()),
+            )),
         }
     }
 
     /// This is the derived public key of the caller given path and predecessor
     /// if predecessor is not provided, it will be the caller of the contract
-    pub fn derived_public_key(&self, path: String, predecessor: Option<AccountId>) -> PublicKey {
+    #[handle_result]
+    pub fn derived_public_key(
+        &self,
+        path: String,
+        predecessor: Option<AccountId>,
+    ) -> Result<PublicKey, MpcContractError> {
         let predecessor = predecessor.unwrap_or_else(env::predecessor_account_id);
         let epsilon = derive_epsilon(&predecessor, &path);
         let derived_public_key =
-            derive_key(near_public_key_to_affine_point(self.public_key()), epsilon);
+            derive_key(near_public_key_to_affine_point(self.public_key()?), epsilon);
         let encoded_point = derived_public_key.to_encoded_point(false);
         let slice: &[u8] = &encoded_point.as_bytes()[1..65];
         let mut data: Vec<u8> = vec![near_sdk::CurveType::SECP256K1 as u8];
         data.extend(slice.to_vec());
-        PublicKey::try_from(data).unwrap()
+        Ok(PublicKey::try_from(data).unwrap())
```

Comment:
we should map this unwrap to an error on the off chance that the derive function doesn't work